[nativert] port semaphore to c10 util

dolpm opened the pull request:
Summary:
nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff adds a simple semaphore interface into c10 until c++20 where we get counting_semaphore

gonna need a oss build export to take a look at this...

Test Plan: CI

Differential Revision: D73882656

Skylion007 commented on c10/util/Semaphore.h:
```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <cstddef>
+
```
Can we have a conditional include / using def that pulls in the C++20 version now, we have some downstream users that already build with C++20

dolpm committed:
[nativert] port semaphore to c10 util (#153504)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/153504

nativert RFC: https://github.com/zhxchen17/rfcs/blob/master/RFC-0043-torch-native-runtime.md

To land the runtime into PyTorch core, we will gradually land logical parts of the code into the Github issue and get each piece properly reviewed.

This diff adds a simple semaphore interface into c10 until c++20 where we get counting_semaphore

gonna need a oss build export to take a look at this...

Test Plan: CI

Reviewed By: zhxchen17

Differential Revision: D73882656
